### PR TITLE
MRPHS-3456: Convert vmware vm to template

### PR DESCRIPTION
**Pull request by avnish30jn**
**Jira-id**

MRPHS-3456

**Problem**

Support to convert a vm to template

**Solution**

Added support to convert vm to template

**Unit Testing** 

Tried converting the powered on and powered off vms,  the vms are converted to template.

### CODE REVIEW

File: virtualmachine/vsphere/util.go [hunk 1]
```diff
@@ -802,7 +802,8 @@ var halt = func(vm *VM) error {
 	vmo := object.NewVirtualMachine(vm.client.Client, vmMo.Reference())
 	poweroffTask, err := vmo.PowerOff(vm.ctx)
 	if err != nil {
-		return fmt.Errorf("error creating a poweroff task on the vm: %s", err)
+		return fmt.Errorf(
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/util.go:815]: is formatting the only change here?
%s->%v

File: virtualmachine/vsphere/vm.go [hunk 2]
```diff
@@ -1587,3 +1587,36 @@ func getVirtualMachines(vm *VM) ([]mo.VirtualMachine, error) {
 	}
 	return vmList, nil
 }
+
+// ConvertToTemplate : converts vm to vm template
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go:1593]: [Queries] I assume vm-template will keep the same name as vm. Please confirm.
Is there a way to rename vm or vm-template later?

**Review by pankajtandon12 — CHANGES_REQUESTED** [review]
Minor comment.

File: virtualmachine/vsphere/util.go (hunk 1 above)
Comment (avnish30jn) [reply virtualmachine/vsphere/util.go:815]: 80 character limit, i was trying to follow this. Trying to correct this wherever i find exceeding limit.

File: virtualmachine/vsphere/vm.go (hunk 2 above)
Comment (avnish30jn) [reply virtualmachine/vsphere/vm.go:1593]: Yes, the name of the template is the same as the vm-name. 
Govmomi supports renaming the vm, but do we need it ?

File: virtualmachine/vsphere/vm.go [hunk 3]
```diff
@@ -1587,3 +1587,36 @@ func getVirtualMachines(vm *VM) ([]mo.VirtualMachine, error) {
...
+	defer vm.cancel()
+
+	dcMo, err := GetDatacenter(vm)
+	if err != nil {
+		return err
+	}
+
+	vmMo, err := findVM(vm, dcMo, vm.Name)
+	if err != nil {
+		return fmt.Errorf("error getting the uploaded VM: %s", err)
```
Comment (pankajtandon12) [inline virtualmachine/vsphere/vm.go]: "%s" -> "%v"
at all places

**Review by pankajtandon12 — CHANGES_REQUESTED** [review]

**Comment by avnish30jn** [comment]
Incorporated PR comments

**Review by pankajtandon12 — APPROVED** [review]
LGTM.

**Pull request merged**